### DISCUSSION TITLE
How to use memray with Gunicorn and Django?

### DISCUSSION THREAD

**Discussion by ghost** [bloomberg/memray#56] [Q&A] [answered]
I have a project with **Django** and I did a multiread with **Gunicorn**. but my project has a memory leak. I want to monitor memory with "**memray**" but I don't know how to use "**memray**".

**Comment by pablogsal** [comment, marked as answer]
Please, check the documentation, especially the `--follow-fork` option if you want to follow forked processes:

https://bloomberg.github.io/memray/getting_started.html

https://bloomberg.github.io/memray/run.html#tracking-across-forks

**Reply by ghost** [reply]
thanks for your responses.

I have two questions:

1- I run Gunicorn and Django with the below command. Why does Memray show one thread in the terminal?

  `memray run --live-remote --live-port 8002 .venv/bin/gunicorn comment_server.wsgi:application -workers=1 --threads=32 `

![Screenshot from 2022-04-25 13-10-15](https://user-images.githubusercontent.com/5820932/165053662-8731eda8-7b36-415f-b884-db3ee87c224c.png)

2- I called  `memray3.8 live` command and tracked memory. after the second I closed the terminal and run the command again but nothing happened and nothing was showing. just show this error in runner `Failed to write output, deactivating tracking`

**Reply by godlygeek** [reply]
> 1- I run Gunicorn and Django with the below command. Why does Memray show one thread in the terminal?

Because the gunicorn arbiter process that you ran under Memray only ever has one thread. It spawns a single worker process (because of `--workers=1`) to actually process the incoming WSGI requests, and that worker process is the one with 32 threads. If you want any transparency into what that worker process is doing, you'll need to use `--follow-fork`, and you won't be able to use live mode (at least not from the command line; it would theoretically be possible to use it from [the Memray API](https://bloomberg.github.io/memray/api.html#memray.SocketDestination)



> 2- I called `memray3.8 live` command and tracked memory. after the second I closed the terminal and run the command again but nothing happened and nothing was showing. just show this error in runner `Failed to write output, deactivating tracking`

It sounds like you're saying you ran `memray --live-remote` once, and then tried to attach to that one running process twice, with two separate calls to `memray live` - do I have that right? If so, that won't work - the live mode server and client need to be synchronized, so the tracked process pauses and waits for a client to connect, and then writes to that client. If that client disconnects, you can't start another - the tracked process is no longer waiting for new connections.

**Reply by ghost** [reply]
Thanks for help 
@godlygeek
